internal/proto/stdout_stderr_writer_test.go

```diff
@@ -0,0 +1,86 @@
+package proto
+
+import (
+	"testing"
+	"bytes"
+	"github.com/stretchr/testify/assert"
+	"fmt"
+	"io"
+	"strings"
+)
+
+func TestOutputStreamSuccess(t *testing.T) {
+	buf := bytes.Buffer{}
+	sendOutput := func(output Output) {
+		fmt.Printf("writing '%v'\n", string(output.Stdout))
+		buf.Write(output.Stdout)
+	}
+	outputStream := MakeStdoutStderrWriter(sendOutput)
+	outputStream.GetStdoutWriter().Write([]byte("hello"))
+	outputStream.GetStdoutWriter().Write([]byte(" world"))
+	outputStream.Close()
+	assert.Equal(t, "hello world", string(buf.Bytes()))
+}
+
+type failingReadCloser struct {
+	name string
+	underlying io.ReadCloser
+	readFails bool
+	closeFails bool
+}
+
+var _ io.ReadCloser = failingReadCloser{}
+
+func (f failingReadCloser) Close() error {
+	if f.closeFails {
+		return fmt.Errorf("closing %v fails!", f.name)
+	} else {
+		return f.underlying.Close()
```

Comment:
can also attach a `*testing.T` to your `failingReadCloser` and `t.Fatal` on an UNexpected error, only return the expected error. then your test can just assert that there's a non-nil err, rather than string-parsing error messages.